op/reboot.go

```diff
@@ -299,143 +628,188 @@ func evictOrDeleteNodePod(ctx context.Context, cs *kubernetes.Clientset, n *cke.
+		switch entry.Status {
+		case cke.RebootStatusDraining, cke.RebootStatusRebooting:
+			alreadyDrained = append(alreadyDrained, entry)
+			if apiServers[entry.Node] {
+				apiServerAlreadyDrained = true
 			}
-			if err != nil {
-				return err
+		case cke.RebootStatusQueued:
+			if entry.DrainBackOffExpire.After(now) {
+				continue
 			}
-			select {
-			case <-ctx.Done():
-				msg := "aborted waiting for pod eviction"
-				log.Error(msg, map[string]interface{}{
-					"namespace": p.Namespace,
-					"name":      p.Name,
-				})
-				return fmt.Errorf("%s: %s/%s", msg, p.Namespace, p.Name)
-			case <-time.After(time.Second * 5):
-				log.Info("waiting for pods to be deleted...", nil)
+			canBeDrained = append(canBeDrained, entry)
+			if apiServerCanBeDrained == nil && apiServers[entry.Node] {
+				apiServerCanBeDrained = entry
 			}
 		}
 	}
-	return nil
+
+	// rules:
+	//   - API Servers are rebooted one by one.
+	//       - It is VERY important.
+	//   - API Servers are rebooted with higher priority than worker nodes.
+	//   - API Servers are not rebooted simultaneously with worker nodes.
+	if apiServerCanBeDrained != nil {
+		if len(alreadyDrained) == 0 {
+			return []*cke.RebootQueueEntry{apiServerCanBeDrained}
+		} else {
+			return nil
+		}
+	}
+	if apiServerAlreadyDrained {
+		return nil
+	}
+	if len(alreadyDrained) >= maxConcurrentReboots {
+		return nil
+	} else if len(alreadyDrained)+len(canBeDrained) <= maxConcurrentReboots {
+		return canBeDrained
+	} else {
+		return canBeDrained[:maxConcurrentReboots-len(alreadyDrained)]
+	}
 }
 
-func (c drainCommand) Run(ctx context.Context, inf cke.Infrastructure, _ string) error {
-	cs, err := inf.K8sClient(ctx, c.apiserver)
+func CheckDrainCompletion(ctx context.Context, inf cke.Infrastructure, apiserver *cke.Node, c *cke.Cluster, rqEntries []*cke.RebootQueueEntry) (completed []*cke.RebootQueueEntry, timedout []*cke.RebootQueueEntry, err error) {
+	evictionTimeoutSeconds := cke.DefaultRebootEvictionTimeoutSeconds
+	if c.Reboot.EvictionTimeoutSeconds != nil {
+		evictionTimeoutSeconds = *c.Reboot.EvictionTimeoutSeconds
+	}
+
+	cs, err := inf.K8sClient(ctx, apiserver)
 	if err != nil {
-		return err
+		return nil, nil, err
 	}
 
-	protected, err := listProtectedNamespaces(ctx, cs, c.protectedNamespaces)
+	protected, err := listProtectedNamespaces(ctx, cs, c.Reboot.ProtectedNamespaces)
 	if err != nil {
-		return err
+		return nil, nil, err
 	}
 
-	for _, n := range c.nodes {
-		err := checkJobPodNotExist(ctx, cs, n)
-		if err != nil {
-			return err
+	t := time.Now().Add(time.Duration(-evictionTimeoutSeconds) * time.Second)
+
+	for _, entry := range rqEntries {
+		if !entry.ClusterMember(c) {
+			continue
+		}
+		if entry.Status != cke.RebootStatusDraining {
+			continue
 		}
-	}
 
-	var targets []*corev1.Pod
-	for _, n := range c.nodes {
-		pods, err := evictOrDeleteNodePod(ctx, cs, n, protected)
-		if err != nil {
-			return err
+		err = checkPodDeletion(ctx, cs, entry.Node, protected)
```

Comment:
This seems somewhat dangerous because a named return value `err` is used.
This `err` may be protected by `err := ...` above, but such a protection is fragile.
I suggest not to use named return values because we return the successful results just at the end.

In general, I prefer to use `err := ` instead of `err = ` in a new block to not overwrite the outside `err`.
`err = ` expresses a will of overwriting even when the value is consumed only in the inner block.
It would be confusing.

Reply:
Agree. I modify it not to use named return value.